File: go/snowflake-id/internal/snowflake/snowflake.go
package snowflake

import (
	"errors"
	"sync"
	"time"
)

const (
	// Layout: 41 bits timestamp (ms) | 10 bits node | 12 bits sequence
	timestampBits = 41
	nodeBits      = 10
	seqBits       = 12

	maxNode = (1 << nodeBits) - 1
	maxSeq  = (1 << seqBits) - 1

	nodeShift = seqBits
	timeShift = nodeBits + seqBits
)

var (
	ErrInvalidNodeID   = errors.New("invalid node id")
	ErrClockWentBack   = errors.New("clock moved backwards")
	ErrEpochInFuture   = errors.New("epoch is in the future")
	defaultCustomEpoch = time.Date(2024, 1, 1, 0, 0, 0, 0, time.UTC)
)

type Generator struct {
	mu    sync.Mutex
	epoch time.Time
	node  uint16

	lastMs int64
	seq    uint16
}

func New(nodeID uint16) (*Generator, error) {
	return NewWithEpoch(nodeID, defaultCustomEpoch)
}

func NewWithEpoch(nodeID uint16, epoch time.Time) (*Generator, error) {
	if nodeID > maxNode {
		return nil, ErrInvalidNodeID
	}
	if epoch.After(time.Now().UTC()) {
		return nil, ErrEpochInFuture
	}
	return &Generator{
		epoch:  epoch.UTC(),
		node:   nodeID,
		lastMs: -1,
	}, nil
}

func (g *Generator) NextID() (uint64, error) {
	g.mu.Lock()
	defer g.mu.Unlock()

	// Measure time in milliseconds from the custom epoch.
	nowMs := time.Since(g.epoch).Milliseconds()
	if nowMs < 0 {
		return 0, ErrEpochInFuture
	}

	// Reject IDs if the observed clock goes backwards.
	if nowMs < g.lastMs {
		return 0, ErrClockWentBack
	}

	if nowMs == g.lastMs {
		// Same millisecond: advance the per-millisecond sequence.
		g.seq = (g.seq + 1) & maxSeq
		if g.seq == 0 {
			// Sequence wrapped, so wait until time moves to the next millisecond.
			for nowMs <= g.lastMs {
				time.Sleep(200 * time.Microsecond)
				nowMs = time.Since(g.epoch).Milliseconds()
			}
		}
	} else {
		// New millisecond: start the sequence from zero again.
		g.seq = 0
	}

	// Remember the latest millisecond we used for the next call.
	g.lastMs = nowMs

	// Keep only the timestamp bits that fit in the Snowflake layout.
	ts := uint64(nowMs) & ((1 << timestampBits) - 1)

	// Pack timestamp, node id, and sequence into one 64-bit ID.
	// Example with small made-up values:
	//   timestamp = 5  -> 101
	//   node      = 3  -> 11
	//   seq       = 2  -> 10
	// After shifting into position:
	//   timestamp << timeShift  puts timestamp in the high bits
	//   uint64(g.node) << nodeShift moves node left by 12 bits, so:
	//   3        = ...000000000011
	//   3 << 12  = ...0011000000000000
	// This leaves the lowest 12 bits free for the sequence value.
	// The final ID is built by OR-ing the three parts together.
	return (ts << timeShift) | (uint64(g.node) << nodeShift) | uint64(g.seq), nil
}

File: django/snowflake-id/internal/snowflake/snowflake.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
import threading
import time


class InvalidNodeID(ValueError):
    pass


class ClockWentBack(RuntimeError):
    pass


class EpochInFuture(RuntimeError):
    pass


_TIMESTAMP_BITS = 41
_NODE_BITS = 10
_SEQ_BITS = 12

_MAX_NODE = (1 << _NODE_BITS) - 1
_MAX_SEQ = (1 << _SEQ_BITS) - 1

_NODE_SHIFT = _SEQ_BITS
_TIME_SHIFT = _NODE_BITS + _SEQ_BITS

_DEFAULT_EPOCH = datetime(2024, 1, 1, 0, 0, 0, 0, tzinfo=timezone.utc)


@dataclass
class Generator:
    node_id: int
    epoch: datetime = _DEFAULT_EPOCH

    def __post_init__(self) -> None:
        if self.node_id < 0 or self.node_id > _MAX_NODE:
            raise InvalidNodeID("invalid node id")
        if self.epoch.tzinfo is None:
            # Keep it simple: force timezone-aware datetimes (UTC).
            self.epoch = self.epoch.replace(tzinfo=timezone.utc)
        self.epoch = self.epoch.astimezone(timezone.utc)

        now = datetime.now(tz=timezone.utc)
        if self.epoch > now:
            raise EpochInFuture("epoch is in the future")

        self._lock = threading.Lock()
        self._last_ms = -1
        self._seq = 0

    def next_id(self) -> int:
        with self._lock:
            now_ms = int((datetime.now(tz=timezone.utc) - self.epoch).total_seconds() * 1000)
            if now_ms < 0:
                raise EpochInFuture("epoch is in the future")

            if now_ms < self._last_ms:
                raise ClockWentBack("clock moved backwards")

            if now_ms == self._last_ms:
                self._seq = (self._seq + 1) & _MAX_SEQ
                if self._seq == 0:
                    # We used up 0..4095 within the same millisecond; wait for the next ms.
                    while now_ms <= self._last_ms:
                        time.sleep(0.0002)  # 200 microseconds
                        now_ms = int((datetime.now(tz=timezone.utc) - self.epoch).total_seconds() * 1000)
            else:
                self._seq = 0

            self._last_ms = now_ms
            ts = now_ms & ((1 << _TIMESTAMP_BITS) - 1)

            return (ts << _TIME_SHIFT) | (self.node_id << _NODE_SHIFT) | self._seq

